### DISCUSSION TITLE
How to make filename the media ID?

### DISCUSSION THREAD

**Discussion by thany** [MalloyDelacroix/DownloaderForReddit#269] [Q&A] [answered]
When I set the `Download Naming` to `%[media_id]` the example promises those IDs we're all familiar with from the previous major version of the program. I want to keep that. But for some reason it's saving files with the post title in their filename, and I don't know how to tell it to use the media ID, other than to literally put it there:

![image](https://user-images.githubusercontent.com/152227/149791296-00fa5d74-76b3-4835-bbd1-52b69ddac65f.png)

It's set like this on the Master User List, as well as on the (only) users list I've created.

A little background:

I have seen it work with `%[post_id]` - it did save with numeric IDs then. But after I found that those weren't the IDs I was hoping to get, I thought I had to reset the user somehow, because downloading the same user again yielding no downloads. So I went into the settings for that one user, and clicked Reset. That didn't cause the same posts to get downloaded again - still yielding 0 downloads. So I decided to just remove the user from the list, and add it again. That worked.

I still didn't have the correct setting though (that's on me), so I quickly terminated the download hoping that it meant I wouldn't have to remove-and-add the user a second time. That made the whole program CTD! So I start it up again, and decided to try this new "Database view" which is super confusing, but I think I managed to delete stuff from there. I thought I was deleting content from a download session, but it turned out it also deleted the users from my users list. I didn't hink it would, because the database view suggested those users to be in a download session.

I guess I just don't understand the whole thing about having that database. I don't get what it's for. But anyhow, I re-added those users, including the one I was testing with, and set it to download again, this time having set the fields like in the screenshot above. And now it's stuck on saving with post title, rather than media ID.

I'm guessing it might be the wrong field. Maybe `media_id` is also the post title or something, or like the title of the actual picture itself. So, what is the correct field to get those `372hj23asd.jpg` kind of files?

Laterly I found that the database view doesn't master-detail from left-to-right, but it feels rather randomish. I have no clue what I'm actually looking at there, and perhaps it's meant more as a debugging tool or something, but I did feel I needed it. Maybe I didn't, and shouldn't have opened it.

Anyway, I digress. So how do I actually get those filenames?
I'm hoping that my assumption, `Download Naming` means `Filename`, is correct? 🤔

**Comment by MalloyDelacroix** [comment, marked as answer]
That is the correct place to set the naming scheme.  I would double check the settings of the user/subreddit(s) you are trying to download and make sure they reflect the same setting.  Could be, for whatever reason, if you changed the setting on the list settings dialog that it did not cascade down to the contained users/subreddits.

In some cases, the "media id" will not be available or of an expected format.  In some instances a website might store the file with the title that the user uploaded it with.  In this case, the "media id" will be the title; although it may not be the same title as the reddit post.  In other instances a "media id" may not be available at all.  In which case a list of other possible ids will be iterated over for a matching value.  So depending on the way that the content is made available on the container website, the "media id" that the content is saved with could be the media id that you are expecting, the id as is stored on reddit, the post title, or even the id as is stored in the application database.

Download naming does mean filename.  As a note, by adding multiple name tokens or custom text to the download naming or save path structure fields and separating them with a "/", you can save content with a deeply nested folder structure.  This will allow you to better organize the content that is downloaded if need be.

**Database concerns:**

In version 3, the app moved to using a database primarily as a more efficient and space saving means of storing data.  In the previous version, all the data that the app used was saved with object storage using the pickle library.  This was inefficient to say the least.  In some cases the save file grew to be several GB in size.  I felt this was unacceptable.  It also lead to persistence issues in which the entire save file would have to be destroyed to regain working order.  Also unacceptable.  An SQLite database is a very efficient and robust solution to storing the data necessary to manage the app.  It also allowed for many more extraction types, extraction/download settings, and several other features to be integrated.  And, since we were going to use a database on the back-end anyway, I figured I might as well take full advantage of it and give users full access to it if they so wanted.

The database view dialog is a user friendly way (I know, calling that dialog user friendly is a bit of a stretch, but bear with me) to view and lightly manage the data stored in the database.  It's more for users who use the app for the content downloading capabilities.  All you can really use the dialog for is to view the data that's been downloaded in a more efficient way than by browsing the download folders and delete some of the items from the database.  It also lets you see _post_ data that would otherwise not be downloaded.

The real power to having everything stored in the database is for users who want to dig deep into analyzing the data.  Since this is a desktop app, the raw database file is fully available to anyone who uses the app.  So someone with more technical skills who is interested in examining reddit data from a data science viewpoint, can access the database directly.  This way they would not have to write their own software to pull data from reddit, and could instead focus on writing software to analyze the data stored in the application database.  This was the vision anyway.

**Reply by thany** [reply]
The way I see it, DFR doesn't need to store anything in a database, except which files have already been downloaded. Okay, that could be called a database of sorts, but it doesn't need a whole (and rather confusing) UI to manage, and cleaning it out could be done fairly automatically.

I myself have built a "chan" downloader. Super simple, super basic. But it too had a downloads database of sorts, which was literally a plaintext file containing every downloaded file on a new line. Even after a million downloaded files, the "database" had grown into 40MB or so. Small enough to load the whole damn thing into memory and be lightning fast. I didn't feel the need to store anything other than file ID's.

The simplest solution is oftenly the best 😀

**Reply by MalloyDelacroix** [reply]
So in your view, the central architecture of the application and almost every one of its features is unnecessary?  That's a bold assessment.  

The database doesn't have to be managed.  You can use the app and never open the database dialog if you don't want to.  But the database is integral to the operation of the entire application.

You may only need the most simplistic features that are offered by the app, and that's fine.  Use only the parts that you want, and ignore the rest.  But some users want and need more advanced features; almost all of which would not be possible by just storing names in text files.

**Reply by thany** [reply]
No, I was setting an example from a personal project. I'm saying if you're going one way or the other, best strive for simplicity rather than complexity.

The thing is, with more advanced features, a more advanced GUI emerges. Sometimes this happens without your or (advanced) users noticing. I'm not saying "go back to basics", but I am saying it's good to keep an eye on this. Try to prevent [feature creep](https://en.wikipedia.org/wiki/Feature_creep).

**Reply by thany** [reply]
On a more practical note, I'm seeing some filenames coming up that I didn't expect:

1. I'm getting some filenames appended with `(1)`. Possibly a way to circumvent duplicate filenames because of the naming scheme I went for. But with Media ID as the filename, this shouldn't be possible. Since those ID's are globally unique on Reddit (if I understand it correctly), getting two files with the same ID means a duplicate download. And indeed, the contents are duplicate. Maybe this is because of crossposting or something. Assuming this isn't a bug, is it possible to prevent these duplicate downloads from happening?
2. I'm also getting filenames like `None 1.jpg` and even they are duplicated by (1), etc., so:
    1. How can a file not have a Media ID? Is this because such file was downloaded from an external site where Media ID isn't a thing?
    2. Is there a way to assign such files an alternate (unique) ID of some sort?
